aws_lambda_powertools/utilities/idempotency/idempotency.py

```diff
@@ -0,0 +1,101 @@
+"""
+Primary interface for idempotent Lambda functions utility
+"""
+import logging
+from typing import Any, Callable, Dict
+
+from aws_lambda_powertools.middleware_factory import lambda_handler_decorator
+from aws_lambda_powertools.utilities.idempotency.persistence.base import STATUS_CONSTANTS, BasePersistenceLayer
+
+from ..typing import LambdaContext
+from .exceptions import AlreadyInProgressError, ItemAlreadyExistsError, ItemNotFoundError
+
+logger = logging.getLogger(__name__)
+
+
+def default_error_callback():
+    raise
+
+
+@lambda_handler_decorator
+def idempotent(
+    handler: Callable[[Any, LambdaContext], Any],
+    event: Dict[str, Any],
+    context: LambdaContext,
+    persistence_store: BasePersistenceLayer,
+) -> Any:
+    """
+    Middleware to handle idempotency
+
+    Parameters
+    ----------
+    handler: Callable
+        Lambda's handler
+    event: Dict
+        Lambda's Event
+    context: Dict
+        Lambda's Context
+    persistence_store: BasePersistenceLayer
+        Instance of BasePersistenceLayer to store data
+
+    Examples
+    --------
+    **Processes Lambda's event in an idempotent manner**
+        >>> from aws_lambda_powertools.utilities.idempotency import idempotent, DynamoDBPersistenceLayer
+        >>>
+        >>> persistence_store = DynamoDBPersistenceLayer(event_key="body", table_name="idempotency_store")
+        >>>
+        >>> @idempotent(persistence_store=persistence_store)
+        >>> def handler(event, context):
+        >>>     return {"StatusCode": 200}
+    """
+
+    try:
+        # We call save_inprogress first as an optimization for the most common case where no idempotent record already
+        # exists. If it succeeds, there's no need to call get_record.
+        persistence_store.save_inprogress(event=event)
+    except ItemAlreadyExistsError:
+        try:
+            event_record = persistence_store.get_record(event)
+        except ItemNotFoundError:
+            return _call_lambda(handler=handler, persistence_store=persistence_store, event=event, context=context)
+
+        if event_record.status == STATUS_CONSTANTS["EXPIRED"]:
+            return _call_lambda(handler=handler, persistence_store=persistence_store, event=event, context=context)
```

Comment:
@cakepietoast possibly log a debug/info message here, as this might be useful to known when debugging expired records

Reply:
@cakepietoast also, should this not be an error?

Reply:
I'll add a log statement. We don't want to raise an error if the idempotency record has expired - we should treat it the same as a missing record and proceed with executing the lambda handler.

Reply:
If it's treated as a missing record, shouldn't the call to save_inprogress happen as in that case?

Reply:
Yep, you're right - we need to retry the save_inprogress call here. This code path would be triggered rarely (when the record isn't expired at initial save_inprogress but is expired at get_record), which would cause a bug. Will fix this.